podbuilder/podbuilder.go

```diff
@@ -0,0 +1,351 @@
+		if bigQueryTable := results.BigQueryTable; bigQueryTable != nil {
+			runContainer.Env = append(runContainer.Env, corev1.EnvVar{
+				Name:  config.BigQueryTableEnv,
+				Value: *bigQueryTable,
+			})
+		}
+	}
+
+	return pod
+}
+
+// PodForServer accepts a pointer to a server and returns a pod for it.
+func (pb *PodBuilder) PodForServer(server *grpcv1.Server) *corev1.Pod {
+	pb.name = safeStrUnwrap(server.Name)
+	pb.role = config.ServerRole
+	pb.pool = safeStrUnwrap(server.Pool)
+	pb.clone = server.Clone
+	pb.build = server.Build
+	pb.run = &server.Run
+
+	pod := pb.newPod()
+
+	runContainer := kubehelpers.ContainerForName(config.RunContainerName, pod.Spec.Containers)
+
+	runContainer.Args = append(runContainer.Args, fmt.Sprintf("--driver_port=%d", pb.defaults.DriverPort))
+	runContainer.Ports = append(runContainer.Ports, corev1.ContainerPort{
+		Name:          "driver",
+		Protocol:      corev1.ProtocolTCP,
+		ContainerPort: pb.defaults.DriverPort,
+	})
+
+	return pod
+}
+
+// newPod creates a base pod for any client, driver or server. It is designed to
+// be decorated by more specific methods for each of these.
+func (pb *PodBuilder) newPod() *corev1.Pod {
+	var initContainers []corev1.Container
+
+	if pb.clone != nil {
+		var env []corev1.EnvVar
+
+		if pb.clone.Repo != nil {
+			env = append(env, corev1.EnvVar{
+				Name:  config.CloneRepoEnv,
+				Value: safeStrUnwrap(pb.clone.Repo),
+			})
+		}
+
+		if pb.clone.GitRef != nil {
+			env = append(env, corev1.EnvVar{
+				Name:  config.CloneGitRefEnv,
+				Value: safeStrUnwrap(pb.clone.GitRef),
+			})
+		}
+
+		initContainers = append(initContainers, corev1.Container{
+			Name:  config.CloneInitContainerName,
+			Image: safeStrUnwrap(pb.clone.Image),
+			Env:   env,
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      config.WorkspaceVolumeName,
+					MountPath: config.WorkspaceMountPath,
+					ReadOnly:  false,
+				},
+			},
+		})
+	}
+
+	if pb.build != nil {
+		initContainers = append(initContainers, corev1.Container{
+			Name:       config.BuildInitContainerName,
+			Image:      safeStrUnwrap(pb.build.Image),
+			Command:    pb.build.Command,
+			Args:       pb.build.Args,
+			Env:        pb.build.Env,
+			WorkingDir: config.WorkspaceMountPath,
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      config.WorkspaceVolumeName,
+					MountPath: config.WorkspaceMountPath,
+					ReadOnly:  false,
+				},
+				{
+					Name:      config.BazelCacheVolumeName,
+					MountPath: config.BazelCacheMountPath,
+					ReadOnly:  false,
+				},
+			},
+		})
+	}
+
+	return &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s-%s", pb.test.Name, pb.role, pb.name),
+			Namespace: pb.test.Namespace,
+			Labels: map[string]string{
+				config.LoadTestLabel:      pb.test.Name,
+				config.RoleLabel:          pb.role,
+				config.ComponentNameLabel: pb.name,
+			},
+		},
+		Spec: corev1.PodSpec{
+			NodeSelector: map[string]string{
```

Comment:
I believe that the logic here could be enhanced to add the "pool" only if it is not empty, and if it is empty, add a "default_pool" corresponding to the pb.role. So the controller configuration could (optionally) specify something like default_worker_pool and default_driver_pool. and that would be added to the "default_pool(s)" map. It would be up to the maintainer of the cluster to ensure that nodes in the cluster are labeled, say, "default_pool=worker", in addition to, say, "pool=worker-8core", in order to make use of this feature.

Reply:
@paulosjca But these default pools should not live in the config/defaults.yaml file, right? After our discussion, would it make sense to have these default pools set in GCP Console where they are maintained? Then, it will try to use those pools. The only problem with defaults is that we do not know if they exist beforehand. It will most likely not produce a clean error, but try repeatedly in the pending state.

Reply:
The enhancement here can be added in a follow-up PR.